docs/language/informal/local-inference.md

```diff
@@ -0,0 +1,498 @@
+there a better formulation of this that is symmetric but still allows some
+propagation?
+
+### Inference rules
+
+- The expression `e as T` is inferred as `m as T` of type `T` in context `K`:
+  - If `e` is inferred as `m` in an empty context
+- The expression `x = e` is inferred as `x = m` of type `T` in context `K`:
+  - If `e` is inferred as `m` of type `T` in context `M` where `x` has type `M`.
+- The expression `x ??= e` is inferred as `x ??= m` of type `UP(T, M)` in
+  context `K`:
+  - If `e` is inferred as `m` of type `T` in context `M` where `x` has type `M`.
+- The expression `await e` is inferred as `await m` of type `T` in context `K`:
+  - If `e` is inferred as `m` of type `T` in context `J` where:
+    - `J` is `FutureOr<K>` if `K` is not `_`, and is `_` otherwise
+- The expression `e0 ?? e1` is inferred as `m0 ?? m1` of type `T` in
+  context `K`:
+  - If `e0` is inferred as `m0` of type `T0` in context `K`
+  - And `e1` is inferred as `m1` of type `T1` in context `J`
+  - Where `J` is `T0` if `K` is `_` and otherwise `K`
+  - Where `T` is the greatest closure of `K` with respect to `?` if `K` is not
+    `_` and otherwise `UP(T0, T1)`
+- The expression `e0..e1` is inferred as `m0..m1` of type `T` in context `K`
+  - If `e0` is inferred as `m0` of type `T` in context `K`
+  - And `e1` is inferred as `m1` of type `P` in context `_`
+- The expression `e0 ? e1 : e2` is inferred as `m0 ? m1 : m2` of type `T` in
+  context `K`
+  - If `e0` is inferred as `m0` of any type in context `bool`
+  - And `e1` is inferred as `m1` of type `T0` in context `K`
+  - And `e2` is inferred as `m2` of type `T1` in context `K`
+  - Where `T` is the greatest closure of `K` with respect to `?` if `K` is not
+    `_` and otherwise `UP(T0, T1)`
+- TODO(leafp): Generalize the following closure cases to the full function
+  signature.
+  - In general, if the function signature is compatible with the context type,
+    take any available information from the context.  If the function signature
+    is not compatible, this this should always be a type error anyway, so
+    implementations should be free to choose the best error recovery path.
+  - The monomorphic case can be treated as a degenerate case of the polymorphic
+    rule
+- The expression `<T>(P x) => e` is inferred as `<T>(P x) => m` of type `<T>(P)
+  -> M` in context `_`
+  - If `e` is inferred as `m` of type `M` in context `_`
+- The expression `<T>(P x) => e` is inferred as `<T>(P x) => m` of type `<T>(P)
+  -> M` in context `<S>(Q) -> N`
+  - If `e` is inferred as `m` of type `M` in context `N[T/S]`
+  - Note: `x` is resolved as having type `P`  for inference in `e`
+- The expression `<T>(x) => e` is inferred as `<T>(dynamic x) => m` of type
+  `<T>(dynamic) -> M` in context `_`
+  - If `e` is inferred as `m` of type `M` in context `_`
+- The expression `<T>(x) => e` is inferred as `<T>(Q[T/S] x) => m` of type
+  `<T>(Q[T/S]) -> M` in context `<S>(Q) -> N`
```

Comment:
All uses of `Q[T/S]` here should be replaced with `R` where `R` is the greatest closure of `Q[T/S]` with respect to `?`.